### ISSUE TITLE
[HOLD for payment 2024-03-22] [HOLD for payment 2024-03-11] [$500] Improve flagging the users as texting Concierge from the OldDot on web

### ISSUE THREAD

**Issue by cristipaval** [Expensify/App#35565]
## Problem
When the user writes to Concierge from the NewDot on the web, we parse the link that started the NewDot session. If that link is a redirect from OldDot (logic [here](https://github.com/Expensify/App/blob/bc4de1278e5126868db75e220971fbb8df0c9404/src/libs/actions/Report.ts#L132-L138) and [here](https://github.com/Expensify/App/blob/bc4de1278e5126868db75e220971fbb8df0c9404/src/libs/actions/Report.ts#L369-L371)), we inform the backend that the user is coming/_texting_ from OldDot.
If the user refreshes the tab, the NewDot session is restarted with the last link accessed at the refresh time. So it happens like this:
1. The user starts Concierge from OldDot web
2. The user gets redirected to the Concierge chat in NewDot in a new tab
3. Sends a message to Concierge
   3.1. in the Concierge tool, the message is marked as coming from OldDot
4. The user refreshes the tab with the Concierge chat in NewDot (this restarts the NewDot session)
5. Sends a new message to Concierge
   5.1 in the Concierge tool, the message is marked as coming from NewDot (though the user is coming from OldDot but for some reason, refreshed the tab)

## Why is this important?
It confuses the agents when they support the users.

## Solution
Make it work properly even if the user refreshes the tab

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e0086d1ab7092595</li>
        <li>Upwork Job ID: 1753082234889191424</li>
        <li>Last Price Increase: 2024-02-15</li>
<li>Automatic offers: </li>
<ul>
<li>abdulrahuman5196 | Reviewer | 0</li>
</ul></li>
    </ul>
</details>

**Comment by MitchExpensify** [comment]
Agree this should be fixed and free submitters makes sense.

**Comment by cristipaval** [comment]
No update yet. I'll spend time on it after I clear up [a CRITICAL](https://github.com/Expensify/App/issues/34982) off my plate.

**Comment by cristipaval** [comment]
no proposals yet Melv, and I'm caught with a CRITICAL issue that blocks comment linking from launching

**Comment by abdulrahuman5196** [comment]
No proposals yet. Should reach out in slack or expert groups to work on this?
@cristipaval / @MitchExpensify

**Comment by cristipaval** [comment]
> No proposals yet. Should we reach out in slack or expert groups to work on this? @cristipaval / @MitchExpensify

great idea, I'll ask an expert group.

**Comment by rezkiy37** [comment]
Hi, I’m Michael (Mykhailo) from Callstack and I would like to work on this issue.

**Comment by cristipaval** [comment]
wooot! Thanks a lot @rezkiy37 🙌

**Comment by rezkiy37** [comment]
Starting to work on the issue on Tuesday.

**Comment by abdulrahuman5196** [comment]
Waiting on PR

**Comment by rezkiy37** [comment]
Starting to work on the issue now.

**Comment by cristipaval** [comment]
Great, thanks for the update! 🙌

**Comment by rezkiy37** [comment]
Btw, how should we inform the support, that I am going to test **Concierge**?

<img width="1235" alt="Screenshot 2024-02-20 at 14 50 55" src="https://github.com/Expensify/App/assets/57314631/713f43f4-ee88-44a1-920d-ba48e3eabab2">

**Comment by cristipaval** [comment]
give me your test account and see if I can assign you to a specific agent

**Comment by rezkiy37** [comment]
I use these 2 accounts: mykhailo.kravchenko@callstack.com and mykhailo.kravchenko+1@callstack.com.

**Comment by cristipaval** [comment]
In the meantime, could we hack the App locally somehow not to perform the API call and watch the console whether the parameter works as we expect?

**Comment by rezkiy37** [comment]
Yes, we can hack.

**Comment by rezkiy37** [comment]
Preparing a proposal for a Callstack's internal review.

**Comment by rezkiy37** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The app informs the backend that the user is coming/texting from OldDot, when the user writes to Concierge on the web. It parses the link that started the NewDot session, if that link is a redirect from OldDot. If the user refreshes the tab, the app loses the initial link from OldDot (_deep-link_). It makes unable to inform the backend that the user texting from the OldDot redirection.

### What is the root cause of that problem?

The app forgets about the initial link (_deep-link_), because it does not persist the value via e.g. **Onyx**. 

### What changes do you think we should make in order to solve the problem?

The idea is to persist the initial url (_deep-link_) via **Onyx** with all params. After the app should take a current value and persisted one, then compare. When there is a new value of `exitTo`, it should use it and store. When there is no new value of `exitTo`, the app uses a previous one, persisted.
Steps: 
1. Add a new key to `ONYXKEYS` - `INITIAL_URL`. The value type is a string.
2. Improve this block of code, where the app assign a value to `reportIDDeeplinkedFromOldDot`.

https://github.com/Expensify/App/blob/20ca041595e76cdd289977e55b2af3fbb9618868/src/libs/actions/Report.ts#L173-L179

  2.1. Connect to the new **Onyx's** key `INITIAL_URL`.
  2.2. Take `reportID` from a current and previous urls.
  2.3. Prefer the current one to the previous one. However, use the second one like a fallback.
  2.4. Set a new initial url, when there is a current `reportID`.

This way the app still can determine that it was redirected from OldDot. Therefore, it sends a correct flag to the backend.

<details><summary>Example how it works</summary>
<p>

https://github.com/Expensify/App/assets/57314631/425e1ec1-9400-42f6-9d20-e71c2109301b

</p>
</details> 

### What alternative solutions did you explore? (Optional) 
N/A

**Comment by cristipaval** [comment]
@rezkiy37 With the above proposal, it sounds like:

1. The user is redirected from OldDot
2. During the session, if the user is texting Concierge, they are marked as coming from OldDot
3. After a refresh, if the user is texting Concierge, they are marked as coming from OldDot
4. The user closes the tab
5. Next day, the user opens a new tab and launches NewDot with the base URL
6. If the user is texting Concierge, they are marked as coming from OldDot, which I don't think is correct.

So the challenge here is to differentiate between a tab refresh and a fresh new session

**Comment by rezkiy37** [comment]
Yes, it works this way. So, if it is incorrect to store the value too long. Let's define some deadlines: tab closing, some time (_1h, 2h and so on_), etc. When must we understand that the user is not coming from OldDot?

**Comment by cristipaval** [comment]
Tab closing sounds great! I'll pull @sonialiap and @flaviadefaria into this as well. So, to summarize:

1. The user opens NewDot by tapping the Concierge button in OldDot
2. During the session, if the user texts Concierge, they are marked as coming from OldDot
3. After a refresh, if the user texts Concierge, they are marked as coming from OldDot
4. The user closes the tab/the browser --> **the persisted information marking the user as coming from OldDot is deleted**
5. Next time, if:
   5.1. the user explicitly opens a new tab and launches NewDot with the base URL -->they are marked as coming from NewDot when they text Concierge
   5.2. the user opens Concierge chat from OldDot again --> they are marked as coming from OldDot when they text Concierge

**Comment by sonialiap** [comment]
@cristipaval your summary looks good. Having the info marking the user as coming from OD if they opened the chat from OD will be very helpful to concierge 🙏

**Comment by rezkiy37** [comment]
Exactly. I will investigate how to track this case when the user closes a tab. In general, I know that the session storage can help for this scenario. Let me please update the proposal.

**Comment by rezkiy37** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The app informs the backend that the user is coming/texting from OldDot, when the user writes to Concierge on the web. It parses the link that started the NewDot session, if that link is a redirect from OldDot. If the user refreshes the tab, the app loses the initial link from OldDot (_deep-link_). It makes unable to inform the backend that the user texting from the OldDot redirection.

### What is the root cause of that problem?

The app forgets about the initial link (_deep-link_), because it does not persist the value via e.g. **Onyx**. 

### What changes do you think we should make in order to solve the problem?

The idea is to persist the initial url (_deep-link_) via the [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage#saving_text_between_refreshes) with all params. After the app should take a current value and persisted one, then compare. When there is a new value of `exitTo`, it should use it and store. When there is no new value of `exitTo`, the app uses a previous one, persisted. The **sessionStorage** satisfies our [requirements](https://github.com/Expensify/App/issues/35565#issuecomment-1959309297) to store the value only when needed for an active tab.

Steps: 
1. Add a new key to `CONST.ts` - `SESSION_STORAGE_KEYS.INITIAL_URL`. The value type is a string.
2. Improve this block of code, where the app assign a value to `reportIDDeeplinkedFromOldDot`.

https://github.com/Expensify/App/blob/20ca041595e76cdd289977e55b2af3fbb9618868/src/libs/actions/Report.ts#L173-L179

  2.1. Check if a current platform is a web one.
  2.2. Get a `prevUrl` from the **sessionStorage**.
  2.3. The same way parse the `prevUrl` as the app does with `url` from **Linking.getInitialURL()**.
  2.4. Prefer the current one to the previous one. However, use the second one like a fallback.
  2.5. Set the new url, when the current one contains a `reportID`.

This way the app still can determine that it was redirected from OldDot. Therefore, it sends a correct flag to the backend.

### What alternative solutions did you explore? (Optional) 
N/A

**Comment by cristipaval** [comment]
The proposal looks great. Just go ahead with it 🚀

**Comment by rezkiy37** [comment]
Good, preparing a PR. Thank you!

**Comment by rezkiy37** [comment]
[The PR](https://github.com/Expensify/App/pull/37088) is being reviewed internally.

**Comment by rezkiy37** [comment]
Going to open the PR for a public review soon.

**Comment by rezkiy37** [comment]
I need to create a follow-up PR to fix a [concern](https://github.com/Expensify/App/pull/37088#discussion_r1511658076).

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

Not a regression.

>Determine if we should create a regression test for this bug.
If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

No. Seems to be an internal feature.

@MitchExpensify

**Comment by MitchExpensify** [comment]
*Payment summary*

C+ $500 @abdulrahuman5196

**Comment by MitchExpensify** [comment]
Paid and contract ended
